src/v/lsm/io/memory_persistence.cc

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "lsm/io/memory_persistence.h"
+
+#include "base/format_to.h"
+#include "lsm/core/exceptions.h"
+
+#include <seastar/core/coroutine.hh>
+
+#include <map>
+#include <memory>
+
+namespace lsm::io {
+
+namespace {
+
+struct memory_file_state {
+    ss::sstring filename;
+    iobuf data;
+    int32_t open_read_handles = 0;
+    int32_t open_write_handles = 0;
+
+    int32_t open_handles() const {
+        return open_read_handles + open_write_handles;
+    }
+};
+
+class memory_sequential_file_reader : public sequential_file_reader {
+public:
+    explicit memory_sequential_file_reader(
+      ss::lw_shared_ptr<memory_file_state> state)
+      : _state(std::move(state)) {
+        if (_state->open_write_handles > 0) {
+            throw io_error_exception(
+              "unable to open new readable file with open write handles: {}",
+              _state->open_handles());
```

Comment:
nit: curious if you found it obvious in handling these exceptions which were the offending files. If not, consider also putting in the filename, here and elsewhere where we throw